pandas/io/excel/_base.py

```diff
@@ -970,21 +1015,39 @@ class ExcelFile:
         "pyxlsb": PyxlsbReader,
     }
 
+    _ext_to_engine: Mapping[str, str] = {
+        ".ods": "odf",
+        ".xls": "xlrd",
+        ".xlsx": "openpyxl",
+    }
+
     def __init__(
         self, path_or_buffer, engine=None, storage_options: StorageOptions = None
     ):
         if engine is None:
-            # Determine ext and use odf for ods stream/file
+
+            ext = peek = None
+
+            if isinstance(path_or_buffer, bytes):
+                path_or_buffer = BytesIO(path_or_buffer)
+
             if isinstance(path_or_buffer, (BufferedIOBase, RawIOBase)):
-                ext = None
-                if _is_ods_stream(path_or_buffer):
-                    engine = "odf"
-            else:
+                engine = _engine_from_content(path_or_buffer)
+                peek = _peek_stream(path_or_buffer)
+
+            elif isinstance(path_or_buffer, (str, os.PathLike)):
                 ext = os.path.splitext(str(path_or_buffer))[-1]
-                if ext == ".ods":
-                    engine = "odf"
+                handles = get_handle(
+                    stringify_path(path_or_buffer),
```

Comment:
Is the `stringigy_path` here needed? (I would assume that `get_handle` handles that)

Reply:
yes, `stringify_path` is not needed for `get_handle`

Reply:
I believe the user expansion that `stringify_path` is needed here.